guardian/amigo
# feat(aws-tools)!: Update AWS CLI to v2

## What does this change?
Install the latest version of the AWS CLI (v2) using an AWS maintained mechanism.

It looks like installing the AWS CLI via `pip3` isn't maintained by AWS as:
- It (still) installs v1
- It is not listed on https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html

Install the AWS CLI via an AWS provided zip file instead.

The package gets installed to `/usr/local/bin/aws`, replicating existing behaviour. This should make this change backwards compatible for places where we're [using the CLI via its full path](https://github.com/search?q=org%3Aguardian+%2Fusr%2Flocal%2Fbin%2Faws++NOT+is%3Aarchived&type=code).

<details><summary>Why not snap?</summary>
<p>

I originally intended to install the CLI via the official `snap` package. However this'll install to `/snap/bin/aws`, and we'd have to symlink this to `/usr/local/bin/aws` for backwards compatibility. This was slightly less readable than installing from the zip file.

</p>
</details> 

## How to test
- Deploy to CODE ([done](https://riffraff.gutools.co.uk/deployment/view/6b8fe730-9965-4e90-8938-93f484611be6))
- Bake a recipe that uses the `aws-tools` role, it should succeed ([done](https://amigo.code.dev-gutools.co.uk/recipes/arm64-jammy-java11-deploy-infrastructure/bakes/27))

## What is the value of this?
- Using the latest version of the AWS CLI.
- This installation mechanism should allow AMIgo to provide better support for Ubuntu 24.04 as when installing via `pip3` we're greeted with an error introduced in Python 3.12:

```log
This environment is externally managed
╰─> To install Python packages system-wide, try apt install
python3-xyz, where xyz is the package you are trying to
install.

If you wish to install a non-Debian-packaged Python package,
create a virtual environment using python3 -m venv path/to/venv.
Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
sure you have python3-full installed.

If you wish to install a non-Debian packaged Python application,
it may be easiest to use pipx install xyz, which will manage a
virtual environment for you. Make sure you have pipx installed.

See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

## Have we considered potential risks?
Version 2 brings some breaking changes - https://docs.aws.amazon.com/cli/latest/userguide/cliv2-migration-changes.html#cliv2-migration-changes-breaking. I can't see any immediate issues this would bring, however I could very well be wrong! If the breaking changes do impact us, I'd expect it to manifest in a failed deployment as the user-data script of an EC2 service would fail to run. That is, there wouldn't be any service disruption.
